src/AlgoliaSearch/FileFailingHostsCache.php

```
@@ -0,0 +1,183 @@
+        }
+    }
+
+    /**
+     * Get failing hosts from cache. This method should also handle cache invalidation if required.
+     * The TTL of the failed hosts cache should be 5mins.
+     *
+     * @return array
+     */
+    public function getFailingHosts()
+    {
+        $cache = $this->loadFailingHostsCacheFromDisk();
+
+        return isset($cache[self::FAILING_HOSTS]) ? $cache[self::FAILING_HOSTS] : array();
+    }
+
+    /**
+     * Removes the file storing the failing hosts.
+     */
+    public function flushFailingHostsCache()
+    {
+        @unlink($this->failingHostsCacheFile);
+    }
+
+    /**
+     * @return array
+     */
+    private function loadFailingHostsCacheFromDisk()
+    {
+        $json = @file_get_contents($this->failingHostsCacheFile);
+        if (false === $json) {
+            return array();
+        }
+
+        $data = json_decode($json, true);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            return array();
+        }
+
+        // Some basic checks.
+        if (
+            !isset($data[self::TTL])
+            || !isset($data[self::FAILING_HOSTS])
+            || !is_int($data[self::TTL])
+            || !is_array($data[self::FAILING_HOSTS])
+        ) {
+            return array();
+        }
+
+        // Validate the hosts array.
+        foreach ($data[self::FAILING_HOSTS] as $host) {
+            if (!is_string($host)) {
+                return array();
+            }
+        }
+
+        $elapsed = time() - $data[self::TTL]; // Number of seconds elapsed.
+
+        if ($elapsed > $this->ttl) {
+            $this->flushFailingHostsCache();
+
+            return array();
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param array $data
+     */
+    private function writeFailingHostsCacheFile(array $data)
+    {
+        $json = json_encode($data);
+        if ($json === false) {
+            return;
+        }
+        @file_put_contents($this->failingHostsCacheFile, $json);
```

Is the `@` really needed? All conditions are checked when the class is created and by suppressing error here we might lose some valuable info for potential debugging.